app/assets/javascripts/stimulus/loaders/autoloader.js

```diff
@@ -3,19 +3,32 @@ import { Application } from "stimulus"
 const application = Application.start()
 
 function autoload() {
-  Array.from(document.querySelectorAll('[data-controller]')).forEach((element) => {
-    const controllerNames = element.attributes["data-controller"].value.split(" ")
+  for (const element of document.querySelectorAll("[data-controller]")) {
+    autoloadControllers(element)
+  }
+}
 
-    controllerNames.forEach((controllerName) => {
-      let controllerFilename = `${controllerName}_controller`
+function autoloadControllers(element) {
+  const controllerTokens = element.getAttribute("data-controller") || ""
+  const controllerNames = controllerTokens.split(/\s+/).filter(content => content.length)
 
-      import(controllerFilename).then((controllerModule) => {
-        application.register(controllerName, controllerModule.default)
-      }).catch(error => console.log(`Failed to autoload controller: ${controllerName}`))
-    })
-  })
+  for (const controllerName of controllerNames) {
+    const filename = controllerName.replace(/--/g, "/").replace(/-/g, "_")
+    import(`${filename}_controller`).then(controllerModule => {
+      application.register(controllerName, controllerModule.default)
+    }).catch(error => console.log(`Failed to autoload controller: ${controllerName}`))
+  }
 }
 
+new MutationObserver((mutationsList) => {
+  mutationsList.forEach(({ attributeName, target }) => {
+    if (attributeName == "data-controller" && target.hasAttribute(attributeName)) {
+      autoloadControllers(target)
+    }
+  })
+}).observe(document.body, { attributeFilter: ["data-controller"], subtree: true, childList: true })
+
 autoload()
 
-window.addEventListener("turbo:load", autoload)
+addEventListener("DOMContentLoaded", autoload)
```

Comment:
The following `addEventListeners` can also be removed because when this code executes it will execute autoload and in all future cases MutationObserver will do loading